### ISSUE TITLE
DPO logits is `nan` from begin when train reasoning dataset with long completion

### ISSUE THREAD

**Issue by AIR-hl** [huggingface/trl#2994]
## command
`accelerate launch --config_file accelerate_config/deepspeed_zero3.yaml dpo.py --config dpo_config.yaml`

both `deepspeed` and `fsdp` have same nan result

## outputs:

![Image](https://github.com/user-attachments/assets/ea7cd371-ee38-482c-867b-ed74eb977a5b)
```
{'loss': 0.6914, 'grad_norm': 5.35485315322876, 'learning_rate': 4.978612153434526e-07, 'rewards/chosen': 0.0, 'rewards/rejected': 0.0, 'rewards/accuracies': 0.0, 'rewards/margins': 0.0, 'logps/chosen': -3513.0, 'logps/rejected': -4191.0, 'logits/chosen': nan, 'logits/rejected': nan, 'epoch': 0.12}
```

## key information
dataset: AIR-hl/OpenR1-Math-220k-paired
model: unsloth/DeepSeek-R1-Distill-Qwen-7B
user_peft: true

## training script
```python
from dataclasses import dataclass
import multiprocessing
import torch
from dataclasses import dataclass, field, asdict, fields
from datasets import load_dataset, Dataset
from tqdm import tqdm
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import (
    DPOConfig,
    DPOTrainer,
    ModelConfig,
    ScriptArguments,
    TrlParser,
    get_kbit_device_map,
    get_peft_config,
    get_quantization_config,
)
from trl.trainer.utils import SIMPLE_CHAT_TEMPLATE

if __name__ == "__main__":
    parser = TrlParser((ScriptArguments, DPOConfig, ModelConfig))
    script_args, training_args, model_config = parser.parse_args_and_config()

    torch_dtype = (
        model_config.torch_dtype
        if model_config.torch_dtype in ["auto", None]
        else getattr(torch, model_config.torch_dtype)
    )

    quantization_config = get_quantization_config(model_config)

    model_kwargs = dict(
        revision=model_config.model_revision,
        attn_implementation=model_config.attn_implementation,
        torch_dtype=torch_dtype,
        use_cache=False if training_args.gradient_checkpointing else True,
        device_map=get_kbit_device_map() if quantization_config is not None else None,
        quantization_config=quantization_config,
    )

    model = AutoModelForCausalLM.from_pretrained(
        model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code, **model_kwargs
    )

    peft_config = get_peft_config(model_config)
    peft_config.target_modules=["gate_proj", "up_proj", "down_proj"]
    
    if peft_config is None:
        ref_model = AutoModelForCausalLM.from_pretrained(
            model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code, **model_kwargs
        )
    else:
        ref_model = None

    tokenizer = AutoTokenizer.from_pretrained(
        model_config.model_name_or_path, trust_remote_code=model_config.trust_remote_code
    )
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    if script_args.ignore_bias_buffers:
        model._ddp_params_and_buffers_to_ignore = [
            name for name, buffer in model.named_buffers() if buffer.dtype == torch.bool
        ]

    tokenizer.chat_template = "{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{% set ns = namespace(system_prompt='') %}{%- for message in messages %}{%- if message['role'] == 'system' %}{% set ns.system_prompt = message['content'] %}{%- endif %}{%- endfor %}{%- if messages|length == 1 and messages[0]['role'] == 'assistant' -%}{%- if messages[0]['content'] is not none -%}{{'<｜Assistant｜>' + messages[0]['content'] + '<｜end▁of▁sentence｜>'}}{%- endif -%}{%- else -%}{{bos_token}}{{ns.system_prompt}}{%- for message in messages %}{%- if message['role'] == 'user' %}{{'<｜User｜>' + message['content']}}{%- endif %}{%- if message['role'] == 'assistant' and message['content'] is not none %}{{'<｜Assistant｜>' + message['content'] + '<｜end▁of▁sentence｜>'}}{%- endif %}{%- endfor -%}{%- endif -%}{% if add_generation_prompt %}{{'<｜Assistant｜>'}}{% endif %}"

    dataset=load_dataset(script_args.dataset_name,
                         split=script_args.dataset_train_split)
    dataset=dataset.select(range(2000))
    dataset=dataset.filter(lambda x: x['chosen_len']>=7172)

    def process(row):
        row['chosen'] = [{'role': 'assistant', 'content': row['chosen']}]
        row['rejected'] = [{'role': 'assistant', 'content': row['rejected']}]
        row['prompt'] = [{'role': 'user', 'content': row['prompt']}]
        return row
    dataset = dataset.map(process, num_proc=multiprocessing.cpu_count())
    
    trainer = DPOTrainer(
        model,
        ref_model,
        args=training_args,
        train_dataset=dataset,
        processing_class=tokenizer,
        peft_config=peft_config,
    )

    train_result = trainer.train()
    trainer.save_model(training_args.output_dir)
```

## dpo_config.yaml
```yaml
model_name_or_path: unsloth/DeepSeek-R1-Distill-Qwen-7B
model_revision: main
torch_dtype: bfloat16
lora_r: 16
lora_alpha: 32
use_peft: true

dataset_name: AIR-hl/OpenR1-Math-220k-paired

bf16: true
max_steps: -1
per_device_train_batch_size: 1
gradient_accumulation_steps: 16
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
learning_rate: 5e-7
lr_scheduler_type: cosine
beta: 0.1
optim: adamw_torch
max_prompt_length: 1024
max_length: 8196

logging_strategy: steps
logging_steps: 1
logging_first_step: true
save_strategy: "steps"
save_steps: 100 
save_total_limit: 4
save_only_model: true
output_dir: test



### System Info

- Platform: Linux-4.19.90-vhulk2211.3.0.h1543.eulerosv2r10.aarch64-aarch64-with-glibc2.28
- Python version: 3.10.6
- PyTorch version: 2.3.1
- CUDA device(s): not available
- Transformers version: 4.49.0
- Accelerate version: 1.4.0
- Accelerate config: not found
- Datasets version: 3.3.2
- HF Hub version: 0.29.1
- TRL version: 0.15.2
- bitsandbytes version: 1.0.0
- DeepSpeed version: 0.16.4
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
DPO is being refactored in #3906, consequently, I'll close this issue. If you think that it should remain open, please let me know
